inc/zoo/swar/SWAR.h

```diff
@@ -83,99 +88,153 @@ namespace detail {
 
     template<bool> struct Selector_impl {
         template<int Level>
-        constexpr static uint64_t execute(uint64_t v) {
+        constexpr static uint64_t execute(uint64_t v) noexcept {
             return popcount_logic<Level>(v);
         }
     };
     template<> struct Selector_impl<true> {
         template<int Level>
-        constexpr static uint64_t execute(uint64_t v) {
+        constexpr static uint64_t execute(uint64_t v) noexcept {
             return popcount_builtin<Level - 2>(v);
         }
     };
 
 }
 
 template<int Level>
-constexpr uint64_t popcount(uint64_t a) {
+constexpr uint64_t popcount(uint64_t a) noexcept {
     return detail::Selector_impl<2 < Level>::template execute<Level>(a);
 }
 
-static_assert(0x210 == popcount<0>(0x320), "");
-static_assert(0x4321 == popcount<1>(0xF754), "");
-static_assert(0x50004 == popcount<3>(0x3E001122), "");
-
 
-template<typename T> constexpr typename std::make_unsigned<T>::type msb_index(T v) {
+template<typename T> constexpr typename std::make_unsigned<T>::type msb_index(T v) noexcept {
     return 8*sizeof(T) - 1 - __builtin_clzll(v);
 }
 
-template<int Size, typename T = uint64_t> struct SWAR {
+template<typename T> constexpr typename std::make_unsigned<T>::type lsb_index(T v) noexcept {
+    return __builtin_ctzll(v) + 1;
+}
+
+template<int NBits, typename T = uint64_t> struct SWAR {
     SWAR() = default;
     constexpr explicit SWAR(T v): m_v(v) {}
+    constexpr explicit operator T() const noexcept { return m_v; }
 
     constexpr T value() const noexcept { return m_v; }
 
-    constexpr SWAR operator|(SWAR o) { return SWAR(m_v | o.m_v); }
-    constexpr SWAR operator&(SWAR o) { return SWAR(m_v & o.m_v); }
-    constexpr SWAR operator^(SWAR o) { return SWAR(m_v ^ o.m_v); }
+    constexpr SWAR operator|(SWAR o) const noexcept { return SWAR(m_v | o.m_v); }
+    constexpr SWAR operator&(SWAR o) const noexcept { return SWAR(m_v & o.m_v); }
+    constexpr SWAR operator^(SWAR o) const noexcept { return SWAR(m_v ^ o.m_v); }
```

Comment:
Damn design of C++ that forces you to do these busy-work things.

I believe this is better with an X-List macro, do you disagree?

Reply:
It is better that way, I just haven't changed it yet.